pkg/server/observer.go

```diff
@@ -289,124 +289,166 @@ func (s *ObserverServer) GetFlows(
 	req *observer.GetFlowsRequest,
 	server observer.Observer_GetFlowsServer,
 ) (err error) {
-	return getFlowsFromObserver(req, server, s)
+	return getFlows(req, server, s)
 }
 
-func getFlowsFromObserver(
+func getFlows(
 	req *observer.GetFlowsRequest,
 	server observer.Observer_GetFlowsServer,
 	obs Observer,
 ) (err error) {
-	reply, err := getFlows(server.Context(), obs.GetLogger(), obs.GetRingBuffer(), req)
+	start := time.Now()
+	log := obs.GetLogger()
+	ring := obs.GetRingBuffer()
+
+	i := uint64(0)
+	defer func() {
+		size := ring.Cap()
+		took := time.Now().Sub(start)
+		log.Debug(
+			"GetFlows finished",
+			zap.Uint64("number_of_flows", i),
+			zap.Uint64("buffer_size", size),
+			zap.String("whitelist", logFilters(req.Whitelist)),
+			zap.String("blacklist", logFilters(req.Blacklist)),
+			zap.Duration("took", took),
+		)
+	}()
+
+	ringReader, err := newRingReader(server.Context(), ring, req)
 	if err != nil {
+		if err == io.EOF {
+			return nil
+		}
 		return err
 	}
-	for {
-		select {
-		case <-server.Context().Done():
-			return nil
-		case rep, ok := <-reply:
-			if !ok {
+	flowsReader, err := newFlowsReader(ringReader, req, log)
+	if err != nil {
+		return err
+	}
+
+	for ; ; i++ {
+		flow, err := flowsReader.Next(server.Context())
+		if err != nil {
+			if err == io.EOF {
 				return nil
 			}
-			err := server.Send(&observer.GetFlowsResponse{
-				ResponseTypes: &observer.GetFlowsResponse_Flow{
-					Flow: rep,
-				},
-			})
-			if err != nil {
-				return err
-			}
+			return err
+		}
+		err = server.Send(&observer.GetFlowsResponse{
+			ResponseTypes: &observer.GetFlowsResponse_Flow{
+				Flow: flow,
+			},
+		})
+		if err != nil {
+			return err
 		}
 	}
 }
 
-func getUntil(req *observer.GetFlowsRequest, defaultTime *types.Timestamp) (time.Time, error) {
-	until := req.GetUntil()
-	if until == nil {
-		until = defaultTime
-	}
-	return types.TimestampFromProto(until)
+type flowsReader struct {
```

Comment:
nesting two "readers" probably warrants a comment who abstracts what

Reply:
Good point, I'll add  doc to `flowsReader`.